cpu/avr8_common/include/cpu.h

```diff
@@ -97,45 +128,25 @@ extern "C"
  * | TX1    | This bit is set when on UART1 TX is pending                   |
  * | TX0    | This bit is set when on UART0 TX is pending                   |
  */
-extern uint8_t avr8_state;
-
-/**
- * @brief   Atomically read the state (@ref avr8_state)
- *
- * This function guarantees that the read is not optimized out, not reordered
- * and done atomically. This does not mean that by the time return value is
- * processed that it still reflects the value currently stored in
- * @ref avr8_state.
- *
- * Using ASM rather than C11 atomics has less overhead, as not every access to
- * the state has to be performed atomically: Those done from ISR will not be
- * interrupted (no support for nested interrupts) and barriers at the begin and
- * end of the ISRs make sure the access takes place before IRQ context is left.
- */
-static inline uint8_t avr8_get_state(void)
-{
-    uint8_t state;
-    __asm__ volatile(
-        "lds   %[state], avr8_state       \n\t"
-        : [state]   "=r" (state)
-        :
-        : "memory"
-
-    );
-
-    return state;
-}
 
 /**
  * @brief   Run this code on entering interrupt routines
  */
 static inline void avr8_enter_isr(void)
 {
-    /* This flag is only called from IRQ context, and nested IRQs are not
-     * supported as of now. The flag will be unset before the IRQ context is
-     * left, so no need to use memory barriers or atomics here
+    /* This flag is only called from IRQ context. The value will be handled
+     * before ISR context is left by @ref avr8_exit_isr.  ATxmega require a
+     * critical section because interrupts are always enabled.
      */
-    avr8_state |= AVR8_STATE_FLAG_ISR;
+#ifdef CPU_ATXMEGA
+    cli();
+#endif
+
+    ++GPIOR0;
```

Comment:
> Note that e.g. on ATmega328P only `GPIOR0` can be used with single-cycle bit-wise instructions.

Could you share where did you get this info?

> An interrupt can only trigger between the instruction 1 and 2, or between 2 and 3. Once the code continues, the contents of r24 (or whatever other register is used) will be restored to the previous state, and so the contents of GPIOR0. For that reason, no critical section is needed here.

I think we need go a little bit dep here. The critical section is to ensure no other interrupt will change content from GPIOR0 and switch context. The above is the code, yes, and it is not an atomic operation at GPIOR0.

This is my readings that I think is important explore, assuming GPIOR0  == 0:
- low IRQ happen
- high regs are pushed to stack
- r24 <- 0
- mid IRQ happen
- high regs are pushed to stack again
- r24 <- 0 (on mid context)  this is wrong! It should be 1 because it is a mid IRQ that interrupts a current in low IRQ.
- r24 = 1
- GPIO0 <- 1
Then system exit from Mid interrupt
- r24 <- 1 (on mid context)
- r24 = 0
- GPIO0 <- 0
- system switch context and stack gets corrupted!
- high regs are poped from stack **from another thread**
- r24 = ?

For ATxmega I believe it is necessary continue with this critical section.

Reply:
> Could you share where did you get this info? There are 3 so, we can pick anyone to do it. But as far I know, all IO registers space allow RMW operations and that is why I would like read this info.

The ATmega328P Datasheet in Section 8.5.1 "General Purpose I/O Registers" says:

> General Purpose I/O Registers within the address range 0x00 - 0x1F are directly bit-accessible using the
SBI, CBI, SBIS, and SBIC instructions.

`GPIOR0` is at 0x1e, `GPIOR1` and `GPIOR2` are at 0x2a and 0x2b, respectively. See also https://godbolt.org/z/WxWdhY9fK

> - low IRQ happen
> - high regs are pushed to stack
> - r24 <- 0
> - mid IRQ happen
> - high regs are pushed to stack again
> - r24 <- 0 (on mid context) this is wrong! It should be 1 because it is a mid IRQ that interrupts a current in low IRQ.

Note that the critical section cannot prevent this. If a mid prio IRQ interrupts a low prio ISR right before the critical section starts, you will still end up reading a `0` from `GPIOR0` despite this already being a nested IRQ.

When the low prio ISR resumes (without critical section), it would restore r24 from the stack. If the IRQ occurred prior to incrementing, r24 and GPIOR0 should again match. The low prio ISR will resume with incrementing r24 and store r24 in GPIOR0. So this should be fine.

Reply:
> GPIOR0 is at 0x1e, GPIOR1 and GPIOR2 are at 0x2a and 0x2b, respectively.

True, I complete forgot about this. ATxmega have first 16 address as GPIORx.

> Note that the critical section cannot prevent this. If a mid prio IRQ interrupts a low prio ISR right before the critical section starts, you will still end up reading a 0 from GPIOR0 despite this already being a nested IRQ.

Agreed. The problem is, an ATxmega mid/high ISR should never switch context when executing **avr8_exit_isr** once stack will corrupt. The way to solve this is checking PMIC.STATUS but code will be slightly different.

Once ATxmega always have ISR enabled (unless using cli/sei) the schedule is secure to perform context switch when STATUS is low. If interrupt STATUS is mid/high it is possible compare and ignore.